File: backend/zvms/views/structs.py
from zvms.routelib import *

Class = Object(
    name=String
)

NoticeBody = Object(
    title=String,
    content=String,
    deadtime=String
)

Notice = Extends(NoticeBody,
    type=Int,
    target=Union(Int, Null)
)

Report = Object(
    report=String
)

Signup = Object(
    volId=Int
)

Login = Object(
    id=Int,
    pwd=String
)

ModPwd = Object(
    old=String,
    new=String
)

ChangeClass = Object(
    cls=Int,
)

Volunteer = Object(
    name=String,
    description=String,
    time=String,
    type=Int,
    reward=Int,
    classes=Array(Object(
        id=Int,
        max=Int
    ))
)

Users = Object(
    users=Array(
        Object(
            id=Int,
            name=String,
            cls=Int,
            auth=Int,
            pwd=String
        )
    )
)

User = Object(
    name=String,
    cls=Int,
    auth=Int
)

File: apis/types/structs.ts
import { int, str, createStructs, arr, createDangerousStructRef, strMaxLength, literal } from "zvms-apis-types-gen";
import { enums } from "./enums.js";

const NoticeBody = {
    title: strMaxLength(32),
    content: strMaxLength(1024),
    deadtime: strMaxLength(19)
}

export const structs = createStructs({
    Class: {
        name: strMaxLength(5)
    },
    NoticeBody,
    Notice: {
        ...NoticeBody,
        type: enums.NoticeType,
        targets: arr(int())
    },
    Report: {
        content: str()
    },
    Signup: {
        stuList: arr(int())
    },
    Thought: {
        content: strMaxLength(1024),
        pictures: arr(str())
    },
    Login: {
        id: int(),
        pwd: strMaxLength(32)
    },
    ModPwd: {
        old: strMaxLength(32),
        new: strMaxLength(32)
    },
    ChangeClass: {
        cls: int()
    },
    UserOfUsers: {
        id: int(),
        name: str(),
        cls: int(),
        auth: int(),
        pwd: str()
    },
    User: {
        name: strMaxLength(5),
        cls: int(),
        auth: int()
    },
    Users: {
        users: arr(createDangerousStructRef("UserOfUsers"))
    },
    VolunteerRecordClass: {
        id: int(),
        max: int()
    },
    Volunteer: {
        name: strMaxLength(32),
        description: strMaxLength(1024),
        time: strMaxLength(20),
        type: literal(enums.VolType.literal.INSIDE,
            enums.VolType.literal.OUTSIDE,
            enums.VolType.literal.LARGE),
        reward: int(),
        classes: arr(createDangerousStructRef("VolunteerRecordClass")),
    },
    Repulse: {
        reason: strMaxLength(1024)
    }
});
